### ISSUE TITLE
Opening up SearchParameters for IndexBinaryHNSW and IndexBinaryIVF

### ISSUE THREAD

**Issue by heemin32** [facebookresearch/faiss#3503]
# Summary

IndexBinaryHNSW and IndexBinaryIVF is blocking SearchParameters. Can we open it up? Will there be any issue that I am not aware of?
https://github.com/facebookresearch/faiss/blob/main/faiss/IndexBinaryHNSW.cpp#L192

**Comment by mdouze** [comment]
what do you mean with "blocking"?

**Comment by heemin32** [comment]
> what do you mean with "blocking"?

```
    FAISS_THROW_IF_NOT_MSG(
            !params, "search params not supported for this index");
```

**Comment by mdouze** [comment]
Right, search parameters are not supported for most binary indexes. Enabling support is relatively simple, feel free to submit a PR.

**Comment by gustavz** [comment]
Hi @mdouze is there an update on this or a PR to track? I am looking for search params support for `IndexBinaryFlat`.
If there is currently no work on it, can you point me to a reference PR that could be used as a base to enable support for that index?

**Comment by heemin32** [comment]
@gustavz You can take a look at this commit as a reference. https://github.com/heemin32/faiss/commit/4d374aa47d4415cbda04b299788988f4ff6e5da0
You can remove following line
```
    if (params_in && params_in->grp) {
        using RH = GroupedHeapBlockResultHandler<HNSW::C>;
        RH bres(n, distances_f, labels, k, params_in->grp);

        hnsw_search(this, n, x, bres, params_in);
    } else {
```

**Comment by gustavz** [comment]
@heemin32 so if thats the code that enables it, why dont you open a PR and get it merged?

**Comment by heemin32** [comment]
@gustavz I was having hard time to build faiss and add unit test in my mac especially the python version of unit test.

**Comment by gustavz** [comment]
@mdouze would you mind making the PR to extend `BinaryIndexFlat` to support search params?
My C skills are unfortunately not enough :/ 
I bet this would be a highly appreciated feature as many use binary indexes for RAG applications and they all want to be able to pre-filter their searches by IDs.

**Comment by gustavz** [comment]
Here is a PR trying to do so, it's mostly cursor generated though: https://github.com/facebookresearch/faiss/pull/4055

**Comment by metonymic-smokey** [comment]
Here's a WIP PR for BIVF indexes - https://github.com/facebookresearch/faiss/pull/4364. 
Will add @mdouze and @asadoughi as reviewers when ready, thanks!

**Comment by gilbsgilbs** [comment]
Hi! I tried to implement ParameterSpace and SearchParameters for Binary HNSW and Binary IVF, including the C APIs in #4761. I safeguarded against ID selection in BinaryIVF for now as there is already work towards this in #4364 and it doesn't look as trivial. Feedback is welcome. Thanks.
